internal/core/graphics/image.rs

```diff
@@ -235,43 +267,92 @@ pub struct StaticTextures {
+    Invalid,
+    /// The image is identified by its path on the file system.
+    Path(SharedString),
+    /// The image is identified by a URL.
+    #[cfg(target_arch = "wasm32")]
+    URL(SharedString),
+    /// The image is identified by the static address of its encoded data.
+    EmbeddedData(usize),
+}
+
+impl ImageCacheKey {
+    /// Returns a new cache key if decoded image data can be stored in image cache for
+    /// the given ImageInner.
+    pub fn new(resource: &ImageInner) -> Option<Self> {
+        let key = match resource {
+            ImageInner::None => return None,
+            ImageInner::EmbeddedImage { cache_key, .. } => cache_key.clone(),
+            ImageInner::StaticTextures(textures) => {
+                Self::from_embedded_image_data(textures.data.as_slice())
+            }
+            #[cfg(feature = "svg")]
+            ImageInner::Svg(parsed_svg) => parsed_svg.cache_key(),
+            #[cfg(target_arch = "wasm32")]
+            ImageInner::HTMLImage(htmlimage) => Self::URL(htmlimage.source().into()),
+        };
+        if matches!(key, ImageCacheKey::Invalid) {
+            None
+        } else {
+            Some(key)
+        }
+    }
+
+    /// Returns a cache key for static embedded image data.
+    pub fn from_embedded_image_data(data: &'static [u8]) -> Self {
+        Self::EmbeddedData(data.as_ptr() as usize)
+    }
+}
+
 /// A resource is a reference to binary data, for example images. They can be accessible on the file
 /// system or embedded in the resulting binary. Or they might be URLs to a web server and a downloaded
 /// is necessary before they can be used.
 /// cbindgen:prefix-with-name
-#[derive(Clone, PartialEq, Debug)]
+#[derive(Clone, Debug)]
 #[repr(u8)]
 #[allow(missing_docs)]
 pub enum ImageInner {
     /// A resource that does not represent any data.
     None,
-    /// A resource that points to a file in the file system
-    AbsoluteFilePath(SharedString),
-    /// A image file that is embedded in the program as is. The format is the extension
-    EmbeddedData {
-        data: Slice<'static, u8>,
-        format: Slice<'static, u8>,
+    EmbeddedImage {
+        cache_key: ImageCacheKey,
+        buffer: SharedImageBuffer,
     },
-    EmbeddedImage(SharedImageBuffer),
+    #[cfg(feature = "svg")]
+    Svg(vtable::VRc<OpaqueRcVTable, svg::ParsedSVG>),
     StaticTextures(&'static StaticTextures),
+    #[cfg(target_arch = "wasm32")]
+    HTMLImage(vtable::VRc<OpaqueRcVTable, htmlimage::HTMLImage>),
 }
 
-impl Default for ImageInner {
-    fn default() -> Self {
-        ImageInner::None
+impl PartialEq for ImageInner {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (
+                Self::EmbeddedImage { cache_key: l_cache_key, buffer: l_buffer },
+                Self::EmbeddedImage { cache_key: r_cache_key, buffer: r_buffer },
+            ) => l_cache_key == r_cache_key && l_buffer == r_buffer,
```

Comment:
Do we need to compare the buffer or is comparing the cahe_key enough?

Reply:
I believe so, yes. Public Image implements PartialEq and wraps ImageInner, so we end up here. You can programmatically crate two Image instances (from_rgba for example) and they will have a ImageCacheKey::Invalid. Therefore IMO we must compare the buffers - they could also be the same if the Image was cloned.